README
# Simpletoolip

...is a Jquery plugin, thought to insert short tooltips to any element of your website more easily

## Features

* Jquery based
* Minimal configuration
* 12 functional positions
* no extra html structures, just use "title" attribute
* cross-browser
* lightweight and fast download, less than 8Kb

## Installation

Download the plugin and decompress files, put the folder simpletooltip in your tree project files, would be nice create a folder that contains it, for example js (/js/simpletooltip).

In the header of your document attach the scripts style-simpletooltip-min.css and jquery-simpletooltip-min.js, of course, you will need load Jquery first:

`<link rel="stylesheet" href="..../jquery-simpletooltip/style-simpletooltip.css" media="screen" />`

`<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.4.4/jquery.min.js"></script>`
`<script src="..../jquery-simpletooltip/jquery-simpletooltip.js"></script>`

Initialize the plugin, it will detect all tooltips marked in the page:

`<script>
    jQuery(document).ready(function() {
        jQuery.simpletooltip();
    });
</script>`


## Usage

You can use Simpletooltip in any element easily, first declare the class simpletooltip, and complete the attribute title, that will be the content of the tooltip, for example:

`<h1 class="simpletooltip" title="And this is his tooltip">_This is a page title</h1>`

Simpletooltip is thought for use small and descriptive texts replacing the original yellow tooltip that browsers offers when we complete the attribute title, Anyway, you can insert content more complex, like linebreaks, lists, images, etc...

`<h1 class="simpletooltip" title="list options:<br /><ul><li>first option</li><li>second option</li><li>third option</li></ul>">lists</h1>`


## Positioning

Simpletooltip has 12 funcional positions, by default is located on 'top' position, you can choose: top, top-left, left-top, left, left-bottom, bottom-left, bottom, bottom-right, right-bottom, right, right-top, top-right.

To add the desired position, add the position name class:

`<div class="simpletooltip right-top">right top</div>`


## Color Themes

Simpletooltip has 4 color themes, the default theme is seablue, you can choose: seablue, pastelblue, darkgray, lightgray.

To add the desired theme, add the theme name class:

`<img src="./images/themesample-seablue.png" class="simpletooltip darkgray" title="Seablue Theme" />`

## Download

You can download the plugin (compressed format) [from here](https://github.com/not-only-code/Simpletooltip/zipball/master), anyway also you can access to GIT repo Contribute with your ideas, new features on Wiki. If this plugin helped you, also you can donate, thank you very much.

changelog.txt
v 1.3.0 (04.03.2013)

+ new: parameters. You can pass some parameters to adjust size, border and colors.
+ new: color schemes, now the default color is black
+ new: added easing functions
+ tweak: arrow images are removed, now are css based
+ tweat: change style system, use .less instead .css
+ tweat: files restructured
+ fix: some bugfixes


v 1.2.0 (01.02.2011)

+ Simpletooltip Logo!
+ new 4 color designs: seablue(default), pastelblue, darkgray, lightgray
+ all new positions (top, top-right, right-top, right, right-bottom, bottom-right, bottom, bottom-left, left-bottom, left, left-top, top-left)
+ new landmark icon design
+ optimized code
+ changed css structure
+ added .js and .css compressed versions
+ scrolling bugs fixed


v 1.1.0 (03.06.2010)

+ 7 new positions ( top , right-top , right , right-bottom , bottom, left-bottom, left, left-top) 
+ icon markup in the target
+ short delay to launch
+ bugs fixed


v 1.0 (19.05.2010)

+ blue-white color version ( with fine shadow box css made)
+ One position: top
+ easy implementation only by css
+ attribute "title" benefits
+ crossbrowser compatible

jquery-simpletooltip.js
/*!
 * SimpleTooltip: show a tooltip in diferents directions from reference
 * carlos.sanz@gmail.com
 * for more info about the plugin please visit: http://github.com/not-only-code/Simpletooltip
**/

(function($) {
	//---- Local debug
	function _debug(msg) {
		if(window.console) {
			console.debug(msg);
		}
	};
	//
	//
	var defaults = {
		direction: 'top',
		theme: '',
		position:{}
	};
	var settings;
	//
	var margins = {
		border: 6,
		top: 15,
		right: 15,
		bottom: 15,
		left: 15
	}
	//
	var themes = ['', 'pastelblue', 'lightgray', 'darkgray'];
	var positions = ['top', 'top-right', 'right-top', 'right', 'right-bottom', 'bottom-right', 'bottom', 'bottom-left', 'left-bottom', 'left', 'left-top', 'top-left'];
	//
	var titles = [];
	//
	//--------- --
	//
	function getThemeName(element) {
		for (var n=0; n < themes.length; n++) {
			if (element.hasClass(themes[n])) return themes[n]
		};
		return settings.theme;
	}
	//--------- --
	//
	function getTooltipPosition(element) {
		for (var n=0; n < positions.length; n++) {
			if (element.hasClass(positions[n])) return positions[n];
		};
		return settings.direction;
	}
	//--------- --
	//
	function getTooltipTag(_event) {
		var title_ = titles[_event.data.index];
		if (title_.length > 0) return('<div id="simple-tooltip-'+_event.data.index+'" class="simple-tooltip '+getTooltipPosition($(_event.target))+' '+getThemeName($(_event.target))+'">'+title_+'<span class="arrow">&nbsp;</span></div>');
		return false;
	}
	//--------- --
	//
	function mouseOver(_event) {
		var element = $(this);
		var tooltip_tag = getTooltipTag(_event);
		if (tooltip_tag) {
			var exist_ = $('body').find('#simple-tooltip-'+_event.data.index);
			if(exist_.length > 0) {
				_event.preventDefault();
				return false;
			}
			var temp_ = $(tooltip_tag).appendTo($('body'));
			temp_.hide();
			reposTooltip(temp_, element);
			temp_.delay(180).fadeIn(200);
		}
		_event.preventDefault();
	};
	//--------- --
	//
	function mouseOut(_event) {
		var exist_ = $('body').find('#simple-tooltip-'+_event.data.index);
		if(exist_.length == 0) {
			_debug('no existe el elemento: cancela');
			_event.preventDefault();
			return false;
		}
		if(exist_.css('opacity') == 0) {
			_debug('opacidad 0: eliminarrrrrr');
			_event.preventDefault();
			exist_.remove();
			return false;
		}
		exist_.clearQueue().stop().fadeOut(100, function(){exist_.remove()});
	};
	//--------- --
	//
	function reposTooltip(tooltip, element) {
		var pos = element.offset();
		//
		switch (getTooltipPosition(element)) {
			default:
			case 'top':
				pos.top -= parseInt(tooltip.outerHeight() + margins.top);
				pos.left += parseInt((element.outerWidth()-tooltip.outerWidth())/2);
			break;
			case 'top-right':
				pos.top -= parseInt(tooltip.outerHeight() + margins.bottom);
				pos.left += parseInt(element.outerWidth() - margins.right - margins.border);
			break;	
			case 'right-top':
				pos.top -= parseInt(tooltip.outerHeight() - margins.bottom);
				pos.left += parseInt(element.outerWidth() + margins.right);
			break;
			case 'right':
				pos.top += parseInt((element.outerHeight() - tooltip.outerHeight())/2);
				pos.left += parseInt(element.outerWidth() + margins.right);
			break;
			case 'right-bottom':
				pos.top += parseInt(element.outerHeight() - margins.bottom);
				pos.left += parseInt(element.outerWidth() + margins.right);
			break;
			case 'bottom-right':
				pos.top += parseInt(element.outerHeight() + margins.bottom);
				pos.left += parseInt(element.outerWidth() - margins.right - margins.border);
			break;
			case 'bottom':
				pos.top += parseInt(element.outerHeight() + margins.bottom);
				pos.left += parseInt((element.outerWidth()-tooltip.outerWidth())/2);
			break;
			case 'bottom-left':
				pos.top += parseInt(element.outerHeight() + margins.bottom);
				pos.left -= parseInt(tooltip.outerWidth() - margins.left - margins.border);
			break;
			case 'left-bottom':
				pos.top += parseInt(element.outerHeight() - margins.bottom);
				pos.left -= parseInt(tooltip.outerWidth() + margins.left);
			break;
			case 'left':
				pos.top += parseInt((element.outerHeight() - tooltip.outerHeight())/2);
				pos.left -= parseInt(tooltip.outerWidth() + margins.left);
			break;
			case 'left-top':
				pos.top -= (tooltip.outerHeight() - margins.bottom);
				pos.left -= parseInt(tooltip.outerWidth() + margins.left);
			break;
			case 'top-left':
				pos.top -= (tooltip.outerHeight() + margins.bottom);
				pos.left -= parseInt(tooltip.outerWidth() - margins.left);
			break;
		}
		//
		tooltip.css('top', pos.top);
		tooltip.css('left', pos.left);
	}
	//--------- --
	//
	function addActions() {
		$('.simpletooltip').each(function(index, value) {
			$(value).css('cursor', 'pointer');
			
			titles[index] = $(value).attr('title');
			$(value).attr('title', '');
			
			$(value).bind('mouseenter', {index:index}, mouseOver);
			$(value).bind('mouseleave', {index:index}, mouseOut);
		})
	};
	//
	function setSettings(settings_) { settings = $.extend(defaults, settings_); };
	//
	//
	$.simpletooltip = function(settings_) {
		//---
		setSettings(settings_);
		//---
		addActions();
	};
})(jQuery);

js/simpletooltip/jquery-simpletooltip-min.js
(function($){var defaults={direction:'top',theme:'',position:{}};var settings;var margins={border:6,top:15,right:15,bottom:15,left:15}
var themes=['','pastelblue','lightgray','darkgray'];var positions=['top','top-right','right-top','right','right-bottom','bottom-right','bottom','bottom-left','left-bottom','left','left-top','top-left'];var titles=[];function getThemeName(element){for(var n=0;n<themes.length;n++){if(element.hasClass(themes[n]))return themes[n]};return settings.theme;}
function getTooltipPosition(element){for(var n=0;n<positions.length;n++){if(element.hasClass(positions[n]))return positions[n];};return settings.direction;}
function getTooltipTag(_event){var title_=titles[_event.data.index];if(title_.length>0)return('<div id="simple-tooltip-'+_event.data.index+'" class="simple-tooltip '+getTooltipPosition($(_event.target))+' '+getThemeName($(_event.target))+'">'+title_+'<span class="arrow"> </span></div>');return false;}
function mouseOver(_event){var element=$(this);var tooltip_tag=getTooltipTag(_event);if(tooltip_tag){var exist_=$('body').find('#simple-tooltip-'+_event.data.index);if(exist_.length>0){_event.preventDefault();return false;}
var temp_=$(tooltip_tag).appendTo($('body'));temp_.hide();reposTooltip(temp_,element);temp_.delay(180).fadeIn(200);}
_event.preventDefault();};function mouseOut(_event){var exist_=$('body').find('#simple-tooltip-'+_event.data.index);if(exist_.length==0){_debug('no existe el elemento: cancela');_event.preventDefault();return false;}
if(exist_.css('opacity')==0){_debug('opacidad 0: eliminarrrrrr');_event.preventDefault();exist_.remove();return false;}
exist_.clearQueue().stop().fadeOut(100,function(){exist_.remove()});};function reposTooltip(tooltip,element){var pos=element.offset();switch(getTooltipPosition(element)){default:case'top':pos.top-=parseInt(tooltip.outerHeight()+margins.top);pos.left+=parseInt((element.outerWidth()-tooltip.outerWidth())/2);break;case'top-right':pos.top-=parseInt(tooltip.outerHeight()+margins.bottom);pos.left+=parseInt(element.outerWidth()-margins.right-margins.border);break;case'right-top':pos.top-=parseInt(tooltip.outerHeight()-margins.bottom);pos.left+=parseInt(element.outerWidth()+margins.right);break;case'right':pos.top+=parseInt((element.outerHeight()-tooltip.outerHeight())/2);pos.left+=parseInt(element.outerWidth()+margins.right);break;case'right-bottom':pos.top+=parseInt(element.outerHeight()-margins.bottom);pos.left+=parseInt(element.outerWidth()+margins.right);break;case'bottom-right':pos.top+=parseInt(element.outerHeight()+margins.bottom);pos.left+=parseInt(element.outerWidth()-margins.right-margins.border);break;case'bottom':pos.top+=parseInt(element.outerHeight()+margins.bottom);pos.left+=parseInt((element.outerWidth()-tooltip.outerWidth())/2);break;case'bottom-left':pos.top+=parseInt(element.outerHeight()+margins.bottom);pos.left-=parseInt(tooltip.outerWidth()-margins.left-margins.border);break;case'left-bottom':pos.top+=parseInt(element.outerHeight()-margins.bottom);pos.left-=parseInt(tooltip.outerWidth()+margins.left);break;case'left':pos.top+=parseInt((element.outerHeight()-tooltip.outerHeight())/2);pos.left-=parseInt(tooltip.outerWidth()+margins.left);break;case'left-top':pos.top-=(tooltip.outerHeight()-margins.bottom);pos.left-=parseInt(tooltip.outerWidth()+margins.left);break;case'top-left':pos.top-=(tooltip.outerHeight()+margins.bottom);pos.left-=parseInt(tooltip.outerWidth()-margins.left);break;}
tooltip.css('top',pos.top);tooltip.css('left',pos.left);}
function addActions(){$('.simpletooltip').each(function(index,value){$(value).css('cursor','pointer');titles[index]=$(value).attr('title');$(value).attr('title','');$(value).bind('mouseenter',{index:index},mouseOver);$(value).bind('mouseleave',{index:index},mouseOut);})};function setSettings(settings_){settings=$.extend(defaults,settings_);};$.simpletooltip=function(settings_){setSettings(settings_);addActions();};})(jQuery);

js/simpletooltip.js
/*!
 * SimpleTooltip: show a tooltip in diferents directions from reference
 * carlos.sanz@gmail.com
 * for more info about the plugin please visit: http://github.com/not-only-code/Simpletooltip
**/

(function($) {
	//---- Local debug
	function _debug (msg) {
		if (window.console) {
			console.debug(msg);
		}
	}
	//--------- --
	//
	var defaults = {
			direction: 'top',
			color: '#cccccc',
			background_color: '#222222',
			border_color: '#111111',
			border_width: 2,
			arrow_width: 6,
			padding: {
				width: 8,
				height: 6
			},
			max_width: 200,
			fade: true,
			position:{}
		},
		settings,
		margins = {
			border: 6,
			top: 15,
			right: 15,
			bottom: 15,
			left: 15
		},
		positions = ['top', 'top-right', 'right-top', 'right', 'right-bottom', 'bottom-right', 'bottom', 'bottom-left', 'left-bottom', 'left', 'left-top', 'top-left'],
		titles = [],
		tooltip_tag = '<div id="simple-tooltip-%index" class="simple-tooltip %position">%title<span class="arrow">&nbsp;</span></div>';
	//
	//--------- --
	//
	function get_tooltip_position (element) {
		for (var n=0; n < positions.length; n++) {
			if (element.hasClass(positions[n])) {
				return positions[n];
			}
		}
		return settings.direction;
	}
	//--------- --
	//
	function get_tooltip_tag (_event) {
		var title = titles[_event.data.index],
			tag;

		if (title.length) {

			tag = tooltip_tag.replace('%index', _event.data.index);
			tag = tag.replace('%position', get_tooltip_position($(_event.target)));
			tag = tag.replace('%title', title);

			return tag;
		}
		return false;
	}
	//--------- --
	//
	function mouse_over (_event) {

		var element = $(this),
			tooltip_id = '#simple-tooltip-' + _event.data.index,
			tag = get_tooltip_tag(_event),
			fade = parse_element_data(element, 'fade');

		if (tag) {

			var exist = $('body').find(tooltip_id);

			if (exist.length) {
				_event.preventDefault();
				return false;
			}

			$(tag).appendTo($('body'));
			var tooltip = $('body').find(tooltip_id);
			tooltip.hide();
			style_tooltip(tooltip, element);

			if (fade) {
				tooltip.delay(180).fadeIn(200);
			} else {
				tooltip.show();
			}
		}
		_event.preventDefault();
	}
	//--------- --
	//
	function mouse_out (_event) {
		_event.preventDefault();

		var tooltip = $('body').find('#simple-tooltip-'+_event.data.index),
			fade = parse_element_data($(this), 'fade');
		//
		if (tooltip.length === 0) {
			return false;
		}
		if (tooltip.css('opacity') === 0) {
			tooltip.remove();
			return false;
		}
		if (fade) {
			tooltip.clearQueue().stop().fadeOut(100, function(){
				tooltip.remove();
			});
		} else {
			tooltip.remove();
		}
	}
	//--------- --
	//
	function parse_element_data (element, data_name) {
		var attribute_var = 'simpletooltip-' + data_name.replace('_', '-');

		if (element.data(attribute_var)) {
			return element.data(attribute_var);
		}

		if (settings[data_name]) {
			return settings[data_name];
		}
		return false;
	}
	//--------- --
	//
	function style_tooltip (tooltip, element) {
		var pos = element.offset(),
			arrow = tooltip.find(' > .arrow'),
			background_color = parse_element_data(element, 'background_color'),
			border_color = parse_element_data(element, 'border_color');
		//
		var border_width = parse_element_data(element, 'border_width');
		border_width = (typeof(border_width) === 'boolean' || border_width === 'none') ? 0 : Number(border_width);
		//
		var arrow_color = (!border_width) ? background_color : border_color;
		//
		var arrow_side_width = Math.round((settings.arrow_width * 3) / 4),
			arrow_position = -parseInt( ((settings.arrow_width * 2) + border_width), 10 ),
			arrow_side_position = -parseInt( ((arrow_side_width * 2) + border_width), 10 );
		//
		tooltip.css({
			maxWidth: parse_element_data(element, 'max_width'),
			backgroundColor: background_color,
			color: parse_element_data(element, 'color'),
			borderColor: border_color,
			borderWidth: border_width
		});
		//
		switch (get_tooltip_position(element)) {
			case 'top-right':
				pos.top -= parseInt( (tooltip.outerHeight() + margins.bottom), 10 );
				pos.left += parseInt( (element.outerWidth() - margins.right - margins.border), 10 );
				arrow.css({
					left: settings.padding.width - border_width,
					borderWidth: arrow_side_width,
					bottom: arrow_side_position,
					borderTopColor: arrow_color,
					borderLeftColor: arrow_color
				});
			break;
			case 'right-top':
				pos.top -= parseInt( (tooltip.outerHeight() - margins.bottom), 10 );
				pos.left += parseInt( (element.outerWidth() + margins.right), 10 );
				arrow.css({
					bottom: settings.padding.height - border_width,
					borderWidth: arrow_side_width,
					left: arrow_side_position,
					borderRightColor: arrow_color,
					borderBottomColor: arrow_color
				});
			break;
			case 'right':
				pos.top += parseInt( ((element.outerHeight() - tooltip.outerHeight())/2), 10 );
				pos.left += parseInt( (element.outerWidth() + margins.right), 10 );
				arrow.css({
					left: arrow_position,
					borderRightColor: arrow_color,
					marginTop: - settings.arrow_width
				});
			break;
			case 'right-bottom':
				pos.top += parseInt( (element.outerHeight() - margins.bottom), 10 );
				pos.left += parseInt( (element.outerWidth() + margins.right), 10 );
				arrow.css({
					top: settings.padding.height - border_width,
					borderWidth: arrow_side_width,
					left: arrow_side_position,
					borderRightColor: arrow_color,
					borderTopColor: arrow_color
				});
			break;
			case 'bottom-right':
				pos.top += parseInt( (element.outerHeight() + margins.bottom), 10 );
				pos.left += parseInt( (element.outerWidth() - margins.right - margins.border), 10 );
				arrow.css({
					left: settings.padding.width - border_width,
					borderWidth: arrow_side_width,
					top: arrow_side_position,
					borderBottomColor: arrow_color,
					borderLeftColor: arrow_color
				});
			break;
			case 'bottom':
				pos.top += parseInt( (element.outerHeight() + margins.bottom), 10 );
				pos.left += parseInt( ((element.outerWidth()-tooltip.outerWidth())/2), 10 );
				arrow.css({
					top: arrow_position,
					marginLeft: - settings.arrow_width,
					borderBottomColor: arrow_color
				});

			break;
			case 'bottom-left':
				pos.top += parseInt( (element.outerHeight() + margins.bottom), 10 );
				pos.left -= parseInt( (tooltip.outerWidth() - margins.left - margins.border), 10 );
				arrow.css({
					right: settings.padding.width - border_width,
					borderWidth: arrow_side_width,
					top: arrow_side_position,
					borderBottomColor: arrow_color,
					borderRightColor: arrow_color
				});
			break;
			case 'left-bottom':
				pos.top += parseInt( (element.outerHeight() - margins.bottom), 10 );
				pos.left -= parseInt( (tooltip.outerWidth() + margins.left), 10 );
				arrow.css({
					top: settings.padding.height - border_width,
					borderWidth: arrow_side_width,
					right: arrow_side_position,
					borderLeftColor: arrow_color,
					borderTopColor: arrow_color
				});
			break;
			case 'left':
				pos.top += parseInt(  ((element.outerHeight() - tooltip.outerHeight())/2), 10 ) ;
				pos.left -= parseInt(  (tooltip.outerWidth() + margins.left), 10 );
				arrow.css({
					right: arrow_position,
					borderLeftColor: arrow_color,
					marginTop: -settings.arrow_width
				});
			break;
			case 'left-top':
				pos.top -= parseInt( (tooltip.outerHeight() - margins.bottom), 10 );
				pos.left -= parseInt( (tooltip.outerWidth() + margins.left), 10 );
				arrow.css({
					bottom: settings.padding.height - border_width,
					borderWidth: arrow_side_width,
					right: arrow_side_position,
					borderLeftColor: arrow_color,
					borderBottomColor: arrow_color
				});
			break;
			case 'top-left':
				pos.top -= parseInt( (tooltip.outerHeight() + margins.bottom), 10 );
				pos.left -= parseInt( (tooltip.outerWidth() - margins.left), 10 );
				arrow.css({
					right: settings.padding.width - border_width,
					borderWidth: arrow_side_width,
					bottom: arrow_side_position,
					borderTopColor: arrow_color,
					borderRightColor: arrow_color
				});
			break;
			// top case
			default:
				pos.top -= parseInt( (tooltip.outerHeight() + margins.top), 10 );
				pos.left += parseInt( ((element.outerWidth()-tooltip.outerWidth())/2), 10 );
				arrow.css({
					bottom: arrow_position,
					borderTopColor: arrow_color,
					marginLeft: - settings.arrow_width
				});
		}
		//
		tooltip.css({
			top: pos.top,
			left: pos.left
		});
	}
	//--------- --
	//
	function add_actions () {
		$('.simpletooltip').each(function(index) {
			$(this).css('cursor', 'pointer');

			titles[index] = $(this).attr('title');
			$(this).attr('title', '');

			$(this).on('mouseenter', {index:index}, mouse_over);
			$(this).on('mouseleave', {index:index}, mouse_out);
		});
	}
	//
	//--------- --
	//
	$.simpletooltip = function(_settings) {
		//---
		settings = $.extend(defaults, _settings);
		_debug(settings);
		//---
		add_actions();
	};
})(jQuery);